Added 'exposed' field to ClassInfo for registered classes

neikeq opened the pull request:
Closes #11913

This is my proposal at fixing the issue I [describe here](https://github.com/godotengine/godot/issues/11913#issuecomment-334967216).

It's a radical change because it forces classes to be registered manually in order to be exposed to scripting languages. However, I think this is the way to go.

I didn't change anything regarding the class reference generation myself, but it should ignore classes that are not exposed.

neikeq commented:
@reduz WDYT of this?

neikeq commented on modules/gdscript/register_types.cpp:
```diff
@@ -42,6 +42,7 @@ void register_gdscript_types() {
 
 	ClassDB::register_class<GDScript>();
 	ClassDB::register_virtual_class<GDFunctionState>();
+	ClassDB::register_virtual_class<GDNativeClass>();
```
Not sure if `GDNativeClass` is meant to be exposed to scripts...

neikeq replied:
@karroffel ?

karroffel replied:
That has nothing to do with me... This is the type of "the core types" in GDScript.

```gdscript
var clazz = Timer # This has type GDNativeClass
var object = clazz.new()
```

So it actually has nothing to do with GDNative :P I don't think it needs to be exposed to scripts though since that's all done in the GDScript module, not the scripts.

neikeq replied:
Ah, right... my mistake. Should we rename GDNative to DLScript to avoid this confusion?

reduz commented:
It seems fine to me

neikeq committed:
Adds 'exposed' field to ClassInfo
This field represents if the class is exposed to the scripting API.
The value is 'true' if the class was registered manually ('ClassDB::register_*class()'), otherwise it's false (registered on '_post_initialize').
- Added missing registration of classes that are meant to be exposed.

neikeq committed:
Mono: Make use of ClassInfo's exposed API
- BindingsGenerator only generates exposed classes.
- Fix creation of managed instances of non-exposed classes.

neikeq commented:
Ready to merge then, unless anyone disagrees.
